shared/src/main/java/com/couchbase/lite/ReadOnlyDocument.java

```diff
@@ -0,0 +1,123 @@
+package com.couchbase.lite;
+
+import com.couchbase.lite.internal.Misc;
+
+
+import java.util.Locale;
+
+/**
+ * Readonly version of the Document.
+ */
+public class ReadOnlyDocument extends ReadOnlyDictionary {
+    //---------------------------------------------
+    // member variables
+    //---------------------------------------------
+    private String id;
+    private CBLC4Doc c4doc;
+
+    //---------------------------------------------
+    // Constructors
+    //---------------------------------------------
+
+    /* package */ ReadOnlyDocument(String id,
+                                   CBLC4Doc c4doc,
+                                   CBLFLDict data) {
+        super(data);
+        this.id = id;
+        this.c4doc = c4doc;
+    }
+
+    //---------------------------------------------
+    // API - public methods
+    //---------------------------------------------
+
+    /**
+     * return the document's ID.
+     *
+     * @return the document's ID
+     */
+    public String getId() {
+        if (id == null)
+            id = Misc.CreateUUID();
```

Comment:
ReadOnlyDocument should not generate a new ID. I understand that when creating a new Document() with null ID, the document will create the ID. Please check.